### PR TITLE
rgw: "reshard cancel" errors with "invalid argument"

### PR DESCRIPTION
**Pull request by ivancich**
When `radosgw-admin reshard cancel ...` was used on a bucket that is
not currently undergoing resharding, it would error out.

Now it will then try to unschedule any reshards for the bucket in the
so-called "reshard log". It will error out if there is neither an
underway resharding nor a scheduled resharding.

Signed-off-by: J. Eric Ivancich <ivancich@redhat.com>

Fixes: https://tracker.ceph.com/issues/53429

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by cbodley** [comment]
@ivancich can you please rebase this on top of the [wip-rgw-multisite-reshard](https://github.com/ceph/ceph/commits/wip-rgw-multisite-reshard) branch?

**Comment by ivancich** [comment]
> @ivancich can you please rebase this on top of the [wip-rgw-multisite-reshard](https://github.com/ceph/ceph/commits/wip-rgw-multisite-reshard) branch?

Sure, @cbodley. But I checked the master code, and cancelling underway resharding can return -EINVAL in RGWBucketReshard::set_resharding_status. So it may be needed in master as well.

**Comment by cbodley** [comment]
> Sure, @cbodley. But I checked the master code, and cancelling underway resharding can return -EINVAL in RGWBucketReshard::set_resharding_status. So it may be needed in master as well.

ah, okay. it's probably easier to take this on master then

**Comment by ivancich** [comment]
@cbodley see: https://github.com/ceph/ceph/pull/44152

**Comment by ivancich** [comment]
jenkins test make check

File: src/rgw/rgw_admin.cc [hunk 1]
```diff
@@ -7329,12 +7334,21 @@ int main(int argc, const char **argv)
...
-    //entry.bucket_id = bucket_id;
 
     ret = reshard.remove(dpp(), entry);
-    if (ret < 0 && ret != -ENOENT) {
-      cerr << "Error in updating reshard log with bucket " <<
-        bucket_name << ": " << cpp_strerror(-ret) << std::endl;
+    if (ret == -ENOENT) {
+      if (!resharding_underway) {
+	cerr << "Error, bucket \"" << bucket_name <<
+	  "\"is neither undergoing resharding nor scheduled to undergo "
```
Comment (cbodley) [inline src/rgw/rgw_admin.cc:7342]: missing space before `is`

Comment (ivancich) [reply src/rgw/rgw_admin.cc:7342]: Thanks, cbodley!

**Review by cbodley — APPROVED** [review]

**Comment by ivancich** [comment]
jenkins retest this please

**Comment by ivancich** [comment]
jenkins test make check

**Comment by cbodley** [comment]
passed qa after rerun:
https://pulpito.ceph.com/cbodley-2022-02-04_00:02:10-rgw-wip-cbodley-testing-distro-default-smithi/
https://pulpito.ceph.com/cbodley-2022-02-04_16:42:47-rgw-wip-cbodley-testing-distro-default-smithi/

**Pull request merged**